pkg/restore/prioritize_group_version_test.go

```diff
@@ -0,0 +1,367 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package restore
+
+import (
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"github.com/vmware-tanzu/velero/pkg/builder"
+	"github.com/vmware-tanzu/velero/pkg/test"
+)
+
+func TestK8sPrioritySort(t *testing.T) {
+	tests := []struct {
+		name string
+		orig []metav1.GroupVersionForDiscovery
+		want []metav1.GroupVersionForDiscovery
+	}{
+		{
+			name: "sorts Kubernetes API group versions per k8s priority",
+			orig: []metav1.GroupVersionForDiscovery{
+				{Version: "v2"},
+				{Version: "v11alpha2"},
+				{Version: "foo10"},
+				{Version: "v10"},
+				{Version: "v12alpha1"},
+				{Version: "v3beta1"},
+				{Version: "foo1"},
+				{Version: "v1"},
+				{Version: "v10beta3"},
+				{Version: "v11beta2"},
+			},
+			want: []metav1.GroupVersionForDiscovery{
+				{Version: "v10"},
+				{Version: "v2"},
+				{Version: "v1"},
+				{Version: "v11beta2"},
+				{Version: "v10beta3"},
+				{Version: "v3beta1"},
+				{Version: "v12alpha1"},
+				{Version: "v11alpha2"},
+				{Version: "foo1"},
+				{Version: "foo10"},
+			},
+		},
+	}
+
+	for _, tc := range tests {
+		t.Run(tc.name, func(t *testing.T) {
+			k8sPrioritySort(tc.orig)
+
+			assert.Equal(t, tc.want, tc.orig)
+		})
+	}
+}
+
+func TestUserResourceGroupVersionPriorities(t *testing.T) {
+	tests := []struct {
+		name string
+		cm   *corev1.ConfigMap
+		want map[string]metav1.APIGroup
+	}{
+		{
+			name: "retrieve version priority data from config map",
+			cm: builder.
+				ForConfigMap("velero", "enableapigroupversions").
+				Data(
+					"restoreResourcesVersionPriority",
+					`rockbands.music.example.io=v2beta1,v2beta2
+orchestras.music.example.io=v2,v3alpha1
+subscriptions.operators.coreos.com=v2,v1`,
+				).
+				Result(),
+			want: map[string]metav1.APIGroup{
+				"rockbands.music.example.io": {Versions: []metav1.GroupVersionForDiscovery{
+					{Version: "v2beta1"},
+					{Version: "v2beta2"},
+				}},
+				"orchestras.music.example.io": {Versions: []metav1.GroupVersionForDiscovery{
+					{Version: "v2"},
+					{Version: "v3alpha1"},
+				}},
+				"subscriptions.operators.coreos.com": {Versions: []metav1.GroupVersionForDiscovery{
+					{Version: "v2"},
+					{Version: "v1"},
+				}},
+			},
+		},
+		{
+			name: "incorrect data format returns no user version priorities",
+			cm: builder.
+				ForConfigMap("velero", "enableapigroupversions").
+				Data(
+					"restoreResourcesVersionPriority",
+					`rockbands.music.example.io=v2beta1,v2beta2\n orchestras.music.example.io=v2,v3alpha1`,
+				).
+				Result(),
+			want: nil,
+		},
+		{
+			name: "incorrect and correct data format returns some user version priorities",
+			cm: builder.
+				ForConfigMap("velero", "enableapigroupversions").
+				Data(
+					"restoreResourcesVersionPriority",
+					`rb=foo,foo2
+o=v2beta2
+\n`,
```

Comment:
suggestion: I think this test case could be improved by using invalid data that's a bit closer to valid data. e.g. `bar=`
Also, you could include some valid lines after the invalid lines to show that it continues processing after an failed validation.

Reply:
I made validation of user provided config maps stricter and now if any lines are invalid, an error will occur. 

It should be easier to see with my changes which are valid lines and which are invalid.